### PR TITLE
Greasemonkey: add script updating

### LINKED ISSUE #1653: bulk updating of greasemonkey scripts
I like the greasemonkey support in qupzilla and use it f.ex. on github. But always checking manually on openuserjs.org for updates is tiresome and errorprone. And to be honest, I only do this when I recognise problems on a website.

Perhaps Qupzilla could check all installed scripts for new versions automatically or upon user request (if you don't want qupzilla to change something behind your back) and install/update the ones selected by the user.

### PR DESCRIPTION
**Pull request by trnkava1**
- Fixed parsing @updateURL/@downloadURL
- Updating is only manual
- User have to click on update button located in Greasemonkey settings
- Update works just like download
- New version is not being checked, original script is replaced

Closes #1653

### CODE REVIEW

File: src/plugins/GreaseMonkey/gm_downloader.h [hunk 1]
```diff
@@ -31,7 +31,10 @@ class GM_Downloader : public QObject
     Q_OBJECT
 public:
     explicit GM_Downloader(const QUrl &url, GM_Manager* manager);
-
+    void updateScript(const QString& fileName);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_downloader.h]: Newline after constructor and before `signals`.

File: src/plugins/GreaseMonkey/gm_downloader.h [hunk 2]
```diff
@@ -31,7 +31,10 @@ class GM_Downloader : public QObject
...
+signals:
+    void finished(const QString& fileName);
+    void error();
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_downloader.h]: Add newline here

File: src/plugins/GreaseMonkey/gm_manager.cpp [hunk 3]
```diff
@@ -256,7 +257,25 @@ void GM_Manager::scriptChanged()
 
 void GM_Manager::doDownloadScript(const QUrl &url)
 {
-    new GM_Downloader(url, this);
+    GM_Downloader* downloader = new GM_Downloader(url, this);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_manager.cpp]: `*` should be at the variable name, so make it `GM_Downloader *downloader`.

File: src/plugins/GreaseMonkey/gm_manager.cpp [hunk 4]
```diff
@@ -256,7 +257,25 @@ void GM_Manager::scriptChanged()
...
+    connect(downloader, &GM_Downloader::finished, this, [=](const QString& filename){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_manager.cpp]: -> `const QString &filename`
Also please change it to `fileName` as you are already using it in other places.

File: src/plugins/GreaseMonkey/gm_manager.cpp [hunk 5]
```diff
@@ -256,7 +257,25 @@ void GM_Manager::scriptChanged()
...
+        bool deleteScript = true;
+        GM_Script* script = new GM_Script(this, filename);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_manager.cpp]: `*` at the var name.

File: src/plugins/GreaseMonkey/gm_script.cpp [hunk 6]
```diff
@@ -148,6 +152,31 @@ QWebEngineScript GM_Script::webScript() const
...
+
+void GM_Script::updateScript()
+{
+    if (!m_downloadUrl.isValid() || m_updating)
+        return;
+
+    m_updating = true;
+    emit updatingChanged(m_updating);
+
+    GM_Downloader* downloader = new GM_Downloader(m_downloadUrl, m_manager);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_script.cpp]: `*` at var

File: src/plugins/GreaseMonkey/gm_script.cpp [hunk 7]
```diff
@@ -148,6 +152,31 @@ QWebEngineScript GM_Script::webScript() const
...
+    downloader->updateScript(m_fileName);
+    connect(downloader, &GM_Downloader::finished, this, [=](){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_script.cpp]: You only need to capture `ŧhis` here, so please change it. Same in the connect bellow.

File: src/plugins/GreaseMonkey/gm_script.h [hunk 8]
```diff
@@ -62,8 +62,11 @@ class GM_Script : public QObject
 
     QWebEngineScript webScript() const;
 
+    bool isUpdating();
+    void updateScript();
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_script.h]: newline here

File: src/plugins/GreaseMonkey/settings/gm_settings.cpp [hunk 9]
```diff
@@ -69,6 +71,16 @@ void GM_Settings::showItemInfo(QListWidgetItem* item)
     dialog->open();
 }
 
+void GM_Settings::updateItem(QListWidgetItem* item)
+{
+    GM_Script* script = getScript(item);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settings.cpp]: `*` at variable

File: src/plugins/GreaseMonkey/settings/gm_settings.cpp [hunk 10]
```diff
@@ -69,6 +71,16 @@ void GM_Settings::showItemInfo(QListWidgetItem* item)
...
+    if (!script) {
+        return;
+    }
+    script->updateScript();
+}
+
+
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settings.cpp]: Extra newline, please remove.

File: src/plugins/GreaseMonkey/settings/gm_settings.cpp [hunk 11]
```diff
@@ -148,6 +160,8 @@ void GM_Settings::loadScripts()
         item->setText(script->name());
         item->setData(Qt::UserRole, script->version());
         item->setData(Qt::UserRole + 1, script->description());
+        item->setData(Qt::UserRole + 2, !script->downloadUrl().isEmpty());
+        item->setData(Qt::UserRole + 3, script->isUpdating());
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settings.cpp]: This really should be made into enum.
But that could be done in another patch.

File: src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp [hunk 12]
```diff
@@ -63,6 +64,8 @@ void GM_SettingsListDelegate::paint(QPainter* painter, const QStyleOptionViewIte
 
     int leftPosition = m_padding;
     int rightPosition = opt.rect.right() - m_padding - 16; // 16 for remove button
+    if(index.data(Qt::UserRole + 2).toBool())
+        rightPosition -=  m_padding + 16; // 16 for update button
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp]: Extra space after `-=`

File: src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp [hunk 13]
```diff
@@ -109,6 +112,16 @@ void GM_SettingsListDelegate::paint(QPainter* painter, const QStyleOptionViewIte
     painter->setFont(opt.font);
     style->drawItemText(painter, infoRect, Qt::TextSingleLine | Qt::AlignLeft, opt.palette, true, info, colorRole);
 
+    // Draw update button
+    if(index.data(Qt::UserRole + 2).toBool()){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp]: Add space after `if` -> `if (index...)`

File: src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp [hunk 14]
```diff
@@ -36,15 +36,17 @@ void GM_SettingsListWidget::mousePressEvent(QMouseEvent* event)
         return;
     }
 
+    if (containsUpdateIcon(event->pos())) {
+        emit updateItemRequested(itemAt(event->pos()));
+        return;
+    }
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp]: add newline

File: src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp [hunk 15]
```diff
@@ -65,3 +67,19 @@ bool GM_SettingsListWidget::containsRemoveIcon(const QPoint &pos) const
 
     return removeIconRect.contains(pos);
 }
+
+bool GM_SettingsListWidget::containsUpdateIcon(const QPoint &pos) const
+{
+    QListWidgetItem* item = itemAt(pos);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp]: `*` at variable.

File: src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp [hunk 16]
```diff
@@ -65,3 +67,19 @@ bool GM_SettingsListWidget::containsRemoveIcon(const QPoint &pos) const
...
+    if (!item || !item->data(Qt::UserRole + 2).toBool())
+        return false;
+
+    const QRect rect = visualItemRect(item);
+    const int updateIconPosition = rect.right() - m_delegate->padding()*2 - 16*2;
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistwidget.cpp]: Please add space: `padding() * 2` and `16 * 2`

**Comment by nowrep** [comment]
It looks great, apart from some coding style issues. Good job!

**Comment by trnkava1** [comment]
Thanks a lot for feedback. I hope I fixed all coding style issues in the second commit 789bb035624835703dbe75c816ad0e202ae91e34

One more thing I forgot to mention - I had big problems with redrawing "update icon" during the script updating. My plan was to make it look like it is disabled but I have not succeed. At least nothing happens when the update is in progress and user clicks again to the icon. Could you help me with that please?

File: src/plugins/GreaseMonkey/gm_downloader.cpp [hunk 17]
```diff
@@ -52,13 +56,15 @@ void GM_Downloader::scriptDownloaded()
         const QByteArray response = QString::fromUtf8(m_reply->readAll()).toUtf8();
 
         if (response.contains(QByteArray("// ==UserScript=="))) {
-            const QString filePath = QString("%1/%2").arg(m_manager->scriptsDirectory(), QzTools::getFileNameFromUrl(m_reply->url()));
-            m_fileName = QzTools::ensureUniqueFilename(filePath);
-
+            if (m_fileName.isEmpty()){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_downloader.cpp]: `if(...) {` - space between `)` and `{`

File: src/plugins/GreaseMonkey/gm_downloader.h [hunk 18]
```diff
@@ -32,6 +32,11 @@ class GM_Downloader : public QObject
 public:
     explicit GM_Downloader(const QUrl &url, GM_Manager* manager);
 
+    void updateScript(const QString& fileName);
+
+signals:
+    void finished(const QString& fileName);
+    void error();
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_downloader.h]: newline

File: src/plugins/GreaseMonkey/gm_manager.cpp [hunk 19]
```diff
@@ -256,7 +257,25 @@ void GM_Manager::scriptChanged()
 
 void GM_Manager::doDownloadScript(const QUrl &url)
 {
-    new GM_Downloader(url, this);
+    GM_Downloader *downloader = new GM_Downloader(url, this);
+    connect(downloader, &GM_Downloader::finished, this, [=](const QString &fileName){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_manager.cpp]: `) {`

File: src/plugins/GreaseMonkey/gm_script.cpp [hunk 20]
```diff
@@ -148,6 +152,31 @@ QWebEngineScript GM_Script::webScript() const
...
+{
+    if (!m_downloadUrl.isValid() || m_updating)
+        return;
+
+    m_updating = true;
+    emit updatingChanged(m_updating);
+
+    GM_Downloader *downloader = new GM_Downloader(m_downloadUrl, m_manager);
+    downloader->updateScript(m_fileName);
+    connect(downloader, &GM_Downloader::finished, this, [this](){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_script.cpp]: `) {`

File: src/plugins/GreaseMonkey/gm_script.cpp [hunk 21]
```diff
@@ -148,6 +152,31 @@ QWebEngineScript GM_Script::webScript() const
...
+        m_updating = false;
+        emit updatingChanged(m_updating);
+    });
+    connect(downloader, &GM_Downloader::error, this, [this](){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/gm_script.cpp]: `) {`

File: src/plugins/GreaseMonkey/settings/gm_settings.cpp [hunk 22]
```diff
@@ -64,11 +66,20 @@ void GM_Settings::showItemInfo(QListWidgetItem* item)
     if (!script) {
         return;
     }
-
-    GM_SettingsScriptInfo* dialog = new GM_SettingsScriptInfo(script, this);
+   GM_SettingsScriptInfo* dialog = new GM_SettingsScriptInfo(script, this);
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settings.cpp]: Fix indent here

File: src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp [hunk 23]
```diff
@@ -63,6 +64,8 @@ void GM_SettingsListDelegate::paint(QPainter* painter, const QStyleOptionViewIte
 
     int leftPosition = m_padding;
     int rightPosition = opt.rect.right() - m_padding - 16; // 16 for remove button
+    if(index.data(Qt::UserRole + 2).toBool())
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp]: `if (`

File: src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp [hunk 24]
```diff
@@ -109,6 +112,16 @@ void GM_SettingsListDelegate::paint(QPainter* painter, const QStyleOptionViewIte
     painter->setFont(opt.font);
     style->drawItemText(painter, infoRect, Qt::TextSingleLine | Qt::AlignLeft, opt.palette, true, info, colorRole);
 
+    // Draw update button
+    if (index.data(Qt::UserRole + 2).toBool()){
```
Comment (nowrep) [inline src/plugins/GreaseMonkey/settings/gm_settingslistdelegate.cpp]: `) {`

**Comment by nowrep** [comment]
> One more thing I forgot to mention - I had big problems with redrawing "update icon" during the script updating. My plan was to make it look like it is disabled but I have not succeed. At least nothing happens when the update is in progress and user clicks again to the icon. Could you help me with that please? 

Yeah, you need to call `QAbstractItemView::update(const QModelIndex &index)` with the index you want to update.

**Comment by nowrep** [comment]
I gave it another look and I can see where the problem is. Repainting of QListWidget is as easy as calling `viewport()->update()`, but it won't work here because the model data is static here. That is, even if you force repaint, the delegate will still pick the old value.

So I think the best way to fix it is to use `GM_Script` directly (as it is already set as `Qt::UserRole + 10`) instead of separate roles for each property.

But this I can fix myself, if you want.

**Comment by trnkava1** [comment]
Thanks again for the feedback. All the coding style issues you mentioned should be fixed now. About that disabling of the button - I think I see where you heading but if you have the idea in your head I would be pleased to have help.

**Comment by nowrep** [comment]
Alright, it looks fine now. Thanks!

I'll merge it now and fix the issue with updating button.

**Comment by trnkava1** [comment]
Ok great. Thanks

**Pull request merged**